### ISSUE TITLE
RFC Centralizing record linking

### ISSUE THREAD

**Issue by kaplun** [inveniosoftware/invenio#2573]
## Current model in Invenio master
Linking among records exist by virtue of dedicated MARC subfields. 

### Citations
E.g. in Invenio master out of the box, this is a reference (as generated by refextract):
```
001329723 999C5 $$0647131$$hC. Quesne and V. M. Tkachuk$$o2$$sJ.Phys.,A37,4267$$y2004
```
`$$0` contains the official citation from record `1329723` to record `647131`. 

This has the following characteristics:
* This implementation exposes the internal record IDs.
* In case a there is no `$0` the other subfields are passed to search engine (by `BibRank` and by `BibFormat`) in order to try to find at least one matching record (which in case would be stored subsequently in `$0`)
* In order to know which records are cited by the current one, it's possible to use the search engine with `999C50:1329723` or use the citation dicts wich are cached by BibRank.
* When a new record is added in the system which happens to be the one referred by a given citation through the other subfields, the new record has to be searched in `999C5s` (using its `773`) in order to see if some potential citation has to updated, in this case:
```
000647131 773__ $$c4267-4281$$pJ.Phys.$$vA37$$y2004
```

### Authors
Authors in `100__` and `700__` have their exact name used to identify the specific MARC field, which is then connected in the `aidPERSONIDPAPERS` table to an author profile.

These author profiles are connected to authority MARC records via e.g. an internal ID such as ORCID.

### Affiliation
In the INSPIRE use case affiliations, next to authors in a paper, are literally connected with a corresponding institute record (via exact string matching).

### Photos/Album
In CDS photos are connected to the parent album in a way that photos metadata are agumented with album metadata at indexing time.

### Other links
Still from the INSPIRE example `773__0` is used to link a book chapter record to the corresponding book record.
`78502w` and `78708w` are used to explain  the superseed/superseeded relation.

## Issues with the current model
* There is not 1 model but several. Each type of link had to implement a dedicated daemon to enforce consistency, each type of link is manipulated in a dedicated way and indexed/cached in a dedicated way
* By explicitly writing in the metadata the recid of the linked record, in case the latter is merged into another one, the latter has to be updated
* It's not possible to express linking with a given level of confidence (e.g. imagine when a link is not made in an authoritative way but only guessed).

# Proposal
Centralize the control of links by having a new table (or SQLAlchemy Model) with the form:

uuid | from | link type | to | confidence
------|------|------|-----|-----
1234-5678-1234-5678 | 123 | refersto | 456 | 0.9
2345-1234-1243-1245 | 234 | authoredby | 567 | 1.0
2345-1234-1243-1245 | 234 | affiliatedat | 678 | 0.5

* The `uuid` would be associated with a given `recjson` field upon ingestion by the uploader workflow, for all those fields that have an outgoing link (for example each author field within a record would receive its own uuid, the same would happen for each reference).
* Ideally, only relations from records representing static entities (e.g. papers) to dynamic one (e.g. people or institutions) should be stored, while the opposite direction should be calculated (and cached) at runtime. In this way we would avoid inconsistencies and continuous swarm of updates.
* The confidence level could be used to store an authoritative link (1.0) or a weaker one (e.g. when it was guessed by a machine learning algorithm or by a fuzzy search). Ideally storing 0 would also have the value of ground truth.
* This table could be used to populate *calculated fields* in recjson, where the recid is stored at runtime when the record is sent to be indexed to ElasticSearch.
* When a relation is coming at input time (e.g. when an incoming paper, states that the author *foo* is really the one of record 123) this relation will be interpreted by a given recjson imported and written into the above relation table (a `uuid` is generated on the fly).
* When a record is modified the `uuid` should be preserved, when the modification is not about changing a given link.
* A new generic widget could be used that curators directly edit specific subset of this table (e.g. all the relations of type `refersto` for record 123. Such a widget could let the curator assign external records for not resolved relations by allowing for autocompletion etc.
* The new bibrank for citation would simply look at this table
* The new bibauthorid would manipulate this table
* The new indexer based on elastic search would consume this table trough recjson
* The new formatter would use this table to create records link

This table might grow over time but, thanks to SQLAlchemy transparent sharding it could be sharded WRT the `link_type` property, thus mapping to several specific tables.

### Side note
Connected with this RFC is the question whether we should still identify records by integers or rather switch to use more generic `uuid` and namespace them e.g. to use the same tool to handle not only bibliographic records but also annotations, comments, documents, library items... This should be the subject of a new RFC.

**Comment by MSusik** [comment]
The work has been moved to #2719 . Please also move the discussion there.

**Comment by jirikuncar** [comment]
>  Please also move the discussion there.

General discussion should stay here. In PR please discuss only code issues. Thanks

**Comment by MSusik** [comment]
Every record in Invenio is a node in a relationship graph (obviously many of them might be isolated). Shouldn't the interface of `Node` class be implemented directly inside `Record` class?

**Comment by MSusik** [comment]
Following yesterday discussion, here are my suggestions:

+ `Neo4j` seems to provide us with all the features that we need. The only problem I see are the persistent identifiers. Every node/edge in `neo4j` graph has a unique id, but this id can be reused by another entity after the given node/edge is deleted. It means we'll have to watch out for any changes and update information in record. In case of using `Neo4j`, the simpliest approach seems to be using `py2neo` and [`flask-neo4j`](https://github.com/lashex/flask-neo4j)
+ There are a lot of open source alternatives to `Neo4j`, and if we want to go this direction, it is worth to discuss all/some of them. Note that `Neo4j` is the most popular solution. Here are few alternatives that I consider the most interesting ones:
  - [`ArangoDB`](https://www.arangodb.com/) - this one I find particularly interesting as it perfectly suits our simpliest use-case. It is a mix of `NOSQL` and `graph` approach. 
  - [`TinkerPop`](http://www.tinkerpop.com/)  - it's actually not an engine, but a standard with own definition of graph model, dataflow framework and query language.  You can check [here](https://github.com/tinkerpop/blueprints/wiki) supported engines. It should be easy to create a solution where the admin of the site decides which engine to use, and we only use `TinkerPop` API to call it.
  - [`Titan`](http://thinkaurelius.github.io/titan/) - various storage backend, quite popular, comparable performance to `neo4j` and _support for `elasticsearch`_ (whatever it means).

There are a lof of engines comparisions (performance-wise and featurewise) available, here I list few of them:
  http://ups.savba.sk/~marek/papers/gdm12-ciglan.pdf
  http://www.stingergraph.com/data/uploads/papers/ppaa2014.pdf
  http://euranova.eu/upl_docs/publications/an-empirical-comparison-of-graph-databases.pdf
  https://docs.google.com/spreadsheet/ccc?key=0AlHPKx74VyC5dERyMHlLQ2lMY3dFQS1JRExYQUNhdVE#gid=0
  http://en.wikipedia.org/wiki/Graph_database
+ One of the arguments raised against our own implementation was that it might be harder to link between different data structures ("models"). As an example the link between `record` and `document` was mentioned. What are the different data structures that might use links? How will we represent `document`?

**Comment by glouppe** [comment]
> One of the arguments raised against our own implementation was that it might be harder to link between different data structures ("models"). As an example the link between record and document was mentioned. What are the different data structures that might use links? How will we represent document?

I was not there yesterday, but for me this also raises the issue: what are the primary citizen objects in Invenio? Do we want to work under the assumption that everything is a record, and therefore that relations should only exists between records? Or do we assume a framework using  a more abstract object representation, eg. SQLAlchemy models, in which case relations would happen only between these objects? Or do we want something else?

Regarding neo4j, what would be the advantages of using an external engine for relations rather than having them as SQLAlchemy objects? I am not concinved the added complexity is worth it. 

In any case, I believe it is time for all of us to finally agree on these core design decisions if we want to move on.

**Comment by jirikuncar** [comment]
### Short Summary

#### Features

- [ ] multiple linkable objects and relationships (`Record` to `Record`, `Record` to `Document`, `Document` to `Document`, ...)
- [ ] simple API (e.g `Relationship.create(Record.get(1), 'citedby', Record.get(2), confidence=0.7, **kwargs)`)
- [ ] helper for calculated JSONAlchemy fields based on relationship type
- [ ] integration with forms - order, relation identifier


#### Implementation

- [ ] start with designing clean API
- [ ] prepare dummy "in memory" implementation to pass tests
- [ ] check possible engines - `SQLAlchemy`, `Neo4J`, ...

**Comment by MSusik** [comment]
> multiple linkable objects and relationships (Record to Record, Record to Document, Document to Document, ...)

Does it mean that I should handle every 'namespace' and it will cover every case?

> check possible engines - SQLAlchemy, Neo4J, ...

Given how the implementation process looks like in Invenio, I can assume that for a long time we will have only one engine available. Can we agree which one to use? I can prepare some comparision for next monday meeting if you want.

**Comment by jirikuncar** [comment]
> Can we agree which one to use?

Start with simple python dict (in memory) to finalize API.

**Comment by tiborsimko** [comment]
Here is promised summary of my comments, most of which we discussed
IRL on Monday.

**WRT** use cases, there are some more, e.g. the linking in Periodicals or
among Institutes, e.g. several other kinds of inter-record linking in
CERN Open Data portal, such as "constituent unit entry", tag 774.  See
[open data JSON MARC configuration](https://github.com/cernopendata/opendata.cern.ch/blob/master/invenio_opendata/base/recordext/fields/opendata.cfg).

**WRT** proposed table, we may need an extra column storing properties of
the relation.  Say record R1 is a photo that was used as B/W on page 7
of record R2, and as RGB on page 2 of record R3.  Where to store this
information?  It logically belongs with the relation, not with the
records themselves.  Hence perhaps a new PostgreSQL JSON column
representing abstract key-value properties store, where people can
store anything they need to describe the properties of the
relationship.  (Kind of like "more info" store in invenio/master.)

(**WRT** various types of relationships, looking at all the various
subfields existing in the MARC relational tags shows that this
property store can become rather big.  Hence perhaps a question
whether an abstract column would be enough or whether we need some
further structure?  I guess a generic JSON column might be enough,
since [it is usually fast](http://simko.home.cern.ch/simko/postgresql-mongodb-json-select-speed.html) to search there, however the plethora of 
relationship properties may require some deeper thoughts.)

**WRT** records vs fields as first-class linkable citizens, wouldn't it be
nice to use UUIDs for almost everything, and allow linking on any part
of JSON nested level.  (See @aw-bib's example of inter-field linking
inside the same MARC record.)  Kind of decorate parts of JSON in a LD
style, meaning "factor this tree out for me into a new record in
another JSON store and link them together".  This would be great for
low-level system users.

(If we make also fields linkable, then this mechanism could be used
for poor-man's authority record control, e.g. field representing
"arXiv subject category" has only ~10 values (hep-th, hep-ph, etc) and
so could just as well live as a link to this "JSON based knowledge
base", so to speak.  When do we use "by-reference linking" and where
do we stop and copy values and use "multi-edit" to manage the linking
information?)

**WRT** possibly representing relation information in record's "master
format" data, we have several use cases: (1) developers or system
librarians are working with JSON all the way as much as possible; (2)
cataloguers and curators might use tools the developers write, or
might prefer to work with "native aster format" such as MARC or
UNIMARC or EAD that is in use on the given concrete Invenio
installation; (3) general public will be served any kind of
"serialised version" of the record, in read-only manner, so this is
less critical.  My concern is about the second category, where we may
need to offer read-write like access to non-JSON formats.

E.g. imagine a site receiving MARC authority records from a third
party; if they want to enrich the information, and send it back, and
if they keep receiving and sending updates to these records, both from
third party and from authors, say, then do we want to represent the
relationship information only in the JSON or also in this "editable
native format" that cataloguers are working with?  If we speak about
an Invenio installation that runs with maybe half an IT person and
five cataloguer persons, using some MS Windows EAD editor or
something, and constantly receiving MARC authority records from the
National Library, the answer would be yes.  Hence, if we represent
linked information in these "native editable formats", we should make
things really lossless, which can go wild.  Would cataloguers agree to
see and edit UUIDs?  Would we hide them semi-transparently somehow?

IOW, would we aim at lossless representation of the relationship
information also in the "native editable format" the concrete site may
use (or not), which can get quite complex just looking at generating
MARC out of the relationship property store?  And if we do, would we
want to represent it in a nice editable manner, not exposing internal
UUIDs too much?

**WRT** using a graph database, this should be definitely seriously
considered...

[raw brain dump]

**Comment by aw-bib** [comment]
@tiborsimko concerning cataloguers to use UUID:

1. I think they will accept them easily
2. but they should not be 40 digits hex codes like git commit ids (IMHO this is not necessary in this usecase)

A real world example where cataloguers handle UUIDs in their native cataloguing interface (`bibedit` in inveniospeak) would e.g. be GBV (the largest union catalogue in Germany; same is true for all other PICA installations) where this is in production for [place a long time her].

What's done, however, is that the editor offers a simple lookup interface to retrieve the UUID easily as outlined above. Also their UUIDs look more like this: 81250898X. IOW: just don't make them to wild ;-) Having source of ids in some other field to make them exchangeable is usually not a problem as well. (`0247_ $2 $a` thingies in [Marc](http://www.loc.gov/marc/bibliographic/bd024.html).) That allows even for quite "lenghty" UUIDs as the final id is formed like `strcat($2, $a)`. In Chris paper guys like [`author:(DE-588b)141792906`](http://d-nb.info/gnd/141792906) ar of this form.

With UUIDs like this cataloguers to my experience are even happy exchange them phone. It was common practice at my job that I got a number like this to refer to a dataset. Note also that their `bibedit` takes care of filling in some fields automagically once the UUID is in place. This helps consistency.

**Comment by MSusik** [comment]
@tiborsimko Thank you for your input!

__WRT__ proposed table, my PR now includes a new column.

__WRT__ records vs fields as first-class linkable citizens. This might drag us into more relational-like approach like the one that is used in `master`. I'm not sure if we want it. The possibility of linking with a single field/subfield of a record is tempting, but then storing data in `JSON` seems not to be the right choice. If we started with storing everything in graph db, then it would be much easier to implement and maintain.  Still:

> "factor this tree out for me into a new record in another JSON store and link them together"

As long as it does not create a physical new record, it is a possibility worth considering.

__WRT__ using a graph database I would suggest discussing it on monday meeting.

**Comment by kaplun** [comment]
After several discussions, parallel meeting, forums... :smile: we came to the current conclusion that relations are going to be implemented in the most naïf way, i.e. by simply storing in a given subfield within the JSON representation of a record (see #2854) the identifier of an other record.

Resolving the external record has to be implemented on a case by case. E.g. elasticsearch will have dedicated enhancer that will borrow subset of pointed records into the linking one to expand the indexing. Jinja will follow links (possibly by instantiating records on the fly?) and caching them.

When a given record is modified a signal will be triggered with the patch of the modification. Record pointing to the modified record will be consequently amended (if necesseray) by following the signal. This should better be detailed in a new RFC.

It is assumed that simple queries related to relations are going to be performed directly through elasticsearch (see #3232). More complex ones will bind into JSONB and depend on PostgreSQL usage.

**Comment by aw-bib** [comment]
@kaplun is `recid` refering to the _record id_ of the instance in question?

This will run into trouble once linked records get interchanged between several instances. (Something we do all the time at join2.)

In your author's use case I may have "Doe, John" as 234 on instance a while it is 12345 on instance b while something more abstract like his orcid would stay the same across instances (as does John Doe).

**Comment by kaplun** [comment]
In the big picture of things you should not imagine you are going to expose exactly the low-level JSON you store inside Invenio as such. When exposing records via API, when sending them to elasticsearch etc. ideally there is going to be some transformation process that will enrich the exposed records (e.g. by borrowing parts of the referred records inline into the referring one)
